ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Basics.hs

```diff
@@ -310,7 +328,7 @@ class ( ShowLedgerState (LedgerTables l)
   -- the 'SingI' constraint. Unfortunately, that is not always the case. The
```

Comment:
```suggestion
  -- the 'IsApplyMapKind' constraint. Unfortunately, that is not always the case. The
```

Reply:
Also I'm not convinced by this comment. `l mk` fails to determine `mk` only when `LedgerTables l` is trivial, in which case `projectLedgerTables` and `withLedgerTables` can be trivially implemented.

Reply:
I _think_ it boils down to the following, though I haven't dotted my is and crossed my ts.

Byron has no ledger tables. Thus a value of type `LedgerTables (LedgerState ByronBlock) mk` gives you know what to determine `mk`.

Cardano always has ledger tables (ie the Shelley UTxO map), even before the hard fork transition from Byron to Shelley.

Our current design involves compositional definitions for `HardForkBlock` which delegate to code on the composed eras (such as `ByronBlock`). Converting a Byron `LedgerTables` to a Cardano `LedgerTables` (eg after applying a block during the Byron era) involves creating an "empty" value of the Shelley UTxO map for whichever `mk` is relevant.

I don't think I've understood the detailed dynamics past that. I would be delighted to remove these constraints.

Reply:
I see now. Permit me to restate to check my understanding:
 the `TableStuff (CardanoBlock c)` instance in `ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs` can't be defined without `polyEmptyLedgerTables`, which certainly requires `IsApplyMapKind mk` for access to `emptyAppliedMK`. 
 
 In my branch I have added an `ApplyMapKind` constructor `ApplyMkName :: String -> ApplyMapKind NameMK k v`. With an open map kind, instead we would define the ver useful `newtype ConstMk a k v = ConsMk a`. 
Clearly We can never `polyEmptyLedgerTables @ (ConstMap a)` for arbitrary a.

Reply:
Yes, perhaps we could replace the `TableStuff CardanoBlock` instance's need for `IsAppliedMapKind` with `HasEmptyMK`. I'd have to spend more time looking into it to be sure.

Reply:
Feel free to try adding such a commit to this branch, if that's reasonable priority for you.

Reply:
I gave it a little try yesterday. The singletons stuff is used by serilalisation/deserialisation and I didn't see an easy way to escape that.